util/pace/util/monitor/netcdf_monitor.py

```diff
@@ -0,0 +1,174 @@
+
+from pace.util.communicator import CubedSphereCommunicator
+
+from .. import _xarray as xr
+from ..filesystem import get_fs
+from ..quantity import Quantity
+
+
+logger = logging.getLogger(__name__)
+
+
+class _TimeChunkedVariable:
+    def __init__(self, initial: Quantity, time_chunk_size: int):
+        self._data = np.zeros(
+            (time_chunk_size, *initial.extent), dtype=initial.data.dtype
+        )
+        self._data[0, ...] = initial.view[:]
+        self._dims = initial.dims
+        self._units = initial.units
+        self._i_time = 1
+
+    def append(self, quantity: Quantity):
+        self._data[self._i_time, ...] = quantity.transpose(self._dims).view[:]
+        self._i_time += 1
+
+    @property
+    def data(self) -> Quantity:
+        return Quantity(
+            data=self._data[: self._i_time, ...],
+            dims=("time",) + tuple(self._dims),
+            units=self._units,
+        )
+
+
+class _ChunkedNetCDFWriter:
+
+    FILENAME_FORMAT = "state_{chunk:04d}.nc"
+
+    def __init__(self, path: str, fs: fsspec.AbstractFileSystem, time_chunk_size: int):
+        self._path = path
+        self._fs = fs
+        self._time_chunk_size = time_chunk_size
+        self._i_time = 0
+        self._chunked: Optional[Dict[str, _TimeChunkedVariable]] = None
+        self._times: List[Any] = []
+        self._time_units: Optional[str] = None
+
+    def append(self, state):
+        logger.debug("appending at time %d", self._i_time)
+        state = {**state}  # copy so we don't mutate the input
+        time = state.pop("time", None)
+        if self._chunked is None:
+            self._chunked = {
+                name: _TimeChunkedVariable(quantity, self._time_chunk_size)
+                for name, quantity in state.items()
+            }
+        else:
+            for name, quantity in state.items():
+                self._chunked[name].append(quantity)
+        self._times.append(time)
+        if (self._i_time + 1) % self._time_chunk_size == 0:
+            logger.debug("flushing on append at time %d", self._i_time)
+            self.flush()
+        self._i_time += 1
+
+    def flush(self):
+        if self._chunked is None:
+            pass
+        else:
+            data_vars = {"time": (["time"], self._times)}
+            for name, chunked in self._chunked.items():
+                data_vars[name] = chunked.data.data_array
+            ds = xr.Dataset(data_vars=data_vars)
+            chunk_index = self._i_time // self._time_chunk_size
+            chunk_path = str(
+                Path(self._path)
+                / _ChunkedNetCDFWriter.FILENAME_FORMAT.format(chunk=chunk_index)
+            )
+            with self._fs.open(chunk_path, "wb") as f:
+                ds.to_netcdf(f)
+
+        self._chunked = None
+        self._times.clear()
+
+
+class NetCDFMonitor:
+    """
+    sympl.Monitor-style object for storing model state dictionaries netCDF files.
+    """
+
+    _CONSTANT_FILENAME = "constants.nc"
```

Comment:
So far the constants store is just lat/lon, I'm hesitant to say change this default to `grid.nc` but I would be confused as to what constants mean. Maybe we can set it from the driver side?

Reply:
I'm hesitant because this adds more coupling between the monitor and the diagnostics - if we create a monitor that says `constants_filename="grid.nc"` then the user has to remember that `self.monitor` assumes all constants are grid data, which is another thing to keep in your head when reading code. I feel like getting confused about this is more likely than not remembering "constants.nc" has grid data after doing an `ncdump -h` once or twice.

If this is too confusing, I can rename the monitor `store_constant` to `store_grid`, and use `grid.nc`. Should I do this?